lib/eventsource/index.js

```diff
@@ -0,0 +1,448 @@
+'use strict'
+
+const { setTimeout } = require('node:timers/promises')
+const { pipeline } = require('node:stream')
+const { mainFetch, Fetch } = require('../fetch')
+const { HeadersList } = require('../fetch/headers')
+const { makeRequest } = require('../fetch/request')
+const { getGlobalOrigin } = require('../fetch/global')
+const { webidl } = require('../fetch/webidl')
+const { CONNECTING, OPEN, CLOSED, defaultReconnectionTime, ANONYMOUS, USE_CREDENTIALS } = require('./constants')
+const { EventSourceStream } = require('./eventsource-stream')
+const { parseMIMEType } = require('../fetch/dataURL')
+const { MessageEvent, ErrorEvent } = require('../websocket/events')
+const { isNetworkError } = require('../fetch/response')
+const { getGlobalDispatcher } = require('../global')
+const { createOpaqueTimingInfo } = require('../fetch/util')
+
+let experimentalWarned = false
+
+/**
+ * @typedef {object} EventSourceInit
+ * @property {boolean} [withCredentials] indicates whether the request
+ * should include credentials.
+ */
+
+/**
+ * The EventSource interface is used to receive server-sent events. It
+ * connects to a server over HTTP and receives events in text/event-stream
+ * format without closing the connection.
+ * @extends {EventTarget}
+ * @see https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events
+ * @api public
+ */
+class EventSource extends EventTarget {
+  #events = {
+    open: null,
+    error: null,
+    message: null
+  }
+
+  #url = null
+  #withCredentials = false
+
+  #readyState = CONNECTING
+
+  #request = null
+  #controller = null
+
+  /**
+   * @type {object}
+   * @property {string} lastEventId
+   * @property {number} reconnectionTime
+   * @property {any} reconnectionTimer
+   */
+  #settings = null
+
+  /**
+   * Creates a new EventSource object.
+   * @param {string} url
+   * @param {EventSourceInit} [eventSourceInitDict]
+   * @see https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface
+   */
+  constructor (url, eventSourceInitDict = {}) {
+    // 1. Let ev be a new EventSource object.
+    super()
+
+    webidl.argumentLengthCheck(arguments, 1, { header: 'EventSource constructor' })
+
+    if (!experimentalWarned) {
+      experimentalWarned = true
+      process.emitWarning('EventSource is experimental, expect them to change at any time.', {
+        code: 'UNDICI-ES'
+      })
+    }
+
+    url = webidl.converters.USVString(url)
+    eventSourceInitDict = webidl.converters.EventSourceInitDict(eventSourceInitDict)
+
+    // 2. Let settings be ev's relevant settings object.
+    this.#settings = {
+      lastEventId: '',
+      reconnectionTime: defaultReconnectionTime
+    }
+
+    // 1. Let baseURL be this's relevant settings object's API base URL.
+    const baseURL = getGlobalOrigin()
+
+    // 2. Let urlRecord be the result of applying the URL parser to url with baseURL.
+    let urlRecord
```

Comment:
step 2 is being done where step 3 is. I don't think you need to do anything special for step 3, so just putting the step 2 comment above it will be fine